File: driver/java/src/main/java/org/controlpaneldriver/serial/PowerMeter.java
package org.controlpaneldriver.serial;

import org.controlpaneldriver.Panel;
import org.controlpaneldriver.exception.AutomationControlError;

import java.util.Map;
import java.util.TreeMap;

import static java.lang.Thread.sleep;
import static org.controlpaneldriver.automation.Registers.Inputs.*;
import static org.controlpaneldriver.modbusutil.ModBusUtil.HOLDING_REGISTER;
import static org.controlpaneldriver.util.Operators.fixedPoint;

public class PowerMeter {

    public static final String KEY_TOTAL_KWH = "totalKWh";
    public static final String KEY_IMPORT_KWH = "importKWh";
    public static final String KEY_EXPORT_KWH = "exportKWh";
    public static final String KEY_ACTIVE_POWER = "activePower";
    public static final String KEY_REACTIVE_POWER = "reactivePower";
    public static final String KEY_REACTIVE_POWER_FAULT = "reactivePowerError";
    public static final String KEY_CURRENT = "current";
    public static final String KEY_VOLTAGE = "voltage";
    public static final String KEY_POWER_FACTOR = "powerFactor";
    public static final String KEY_FREQUENCY = "frequency";

    private final Panel panel;

    private int totalKWh = 0;
    private int importKwh = 0;
    private int exportKWh = 0;
    private int reactivePower = 0;
    private short activePower = 0;
    private short current = 0;
    private short voltage = 0;
    private short powerFactor = 0;
    private short frequency = 0;

    private boolean eventChanged = true;

    private boolean totalKWhUpdated = true;
    private boolean importKwhUpdated = true;
    private boolean exportKWhUpdated = true;
    private boolean reactivePowerUpdated = true;
    private boolean activePowerUpdated = true;
    private boolean currentUpdated = true;
    private boolean voltageUpdated = true;
    private boolean powerFactorUpdated = true;
    private boolean frequencyUpdated = true;

    private TreeMap<String, Object> sensorResult = null;
    private TreeMap<String, Object> sensorResultStr = null;

    private void powerMeterCleanUp() {
        this.totalKWh = 0;
        this.importKwh = 0;
        this.exportKWh = 0;
        this.reactivePower = 0;
        this.activePower = 0;
        this.current = 0;
        this.voltage = 0;
        this.powerFactor = 0;
        this.frequency = 0;

        this.eventChanged = true;

        this.totalKWhUpdated = true;
        this.importKwhUpdated = true;
        this.exportKWhUpdated = true;
        this.reactivePowerUpdated = true;
        this.activePowerUpdated = true;
        this.currentUpdated = true;
        this.voltageUpdated = true;
        this.powerFactorUpdated = true;
        this.frequencyUpdated = true;

        if (this.sensorResult != null)
            this.sensorResult.clear();

        if (this.sensorResultStr != null)
            this.sensorResultStr.clear();

        panel.unlock();
    }
    private void updateHelper(Map<String, Object> destination, String key, Object value, boolean shouldAdd) {
        if (shouldAdd)
            destination.put(key, value);
    }

    public PowerMeter(Panel panel) {
        this.panel = panel;
    }

    public void readSensors(boolean forceUpdate) throws Throwable {
        int intValue;
        short shortValue;

        while (panel.isLocked())
            sleep(10);

        panel.lock();

        try {
            intValue = panel.readInt(TOTAL_KWH_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.totalKWhUpdated = (intValue != this.totalKWh);
        this.eventChanged = this.totalKWhUpdated;
        this.totalKWh = intValue;

        try {
            intValue = panel.readInt(IMPORT_KWH_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.importKwhUpdated = (intValue != this.importKwh);
        this.eventChanged |= this.importKwhUpdated;
        this.importKwh = intValue;

        try {
            intValue = panel.readInt(EXPORT_KWH_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.exportKWhUpdated = (intValue != this.exportKWh);
        this.eventChanged |= this.exportKWhUpdated;
        this.exportKWh = intValue;

        try {
            shortValue = panel.readShort(ACTIVE_POWER_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.activePowerUpdated = (shortValue != this.activePower);
        this.eventChanged |= this.activePowerUpdated;
        this.activePower = shortValue;

        try {
            shortValue = panel.readShort(CURRENT_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.currentUpdated = (shortValue != this.current);
        this.eventChanged |= this.currentUpdated;
        this.current = shortValue;

        try {
            shortValue = panel.readShort(VOLTAGE_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.voltageUpdated = (shortValue != this.voltage);
        this.eventChanged |= this.voltageUpdated;
        this.voltage = shortValue;

        try {
            shortValue = panel.readShort(POWER_FACTOR_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.powerFactorUpdated = (shortValue != this.powerFactor);
        this.eventChanged |= this.powerFactorUpdated;
        this.powerFactor = shortValue;

        try {
            shortValue = panel.readShort(FREQUENCY_REG, HOLDING_REGISTER);
        } catch (Throwable e) {
            powerMeterCleanUp();
            throw e;
        }

        this.frequencyUpdated = (shortValue != this.frequency);
        this.eventChanged |= this.frequencyUpdated;
        this.frequency = shortValue;

        try {
            intValue = getReactivePowerHelper();
        } catch (AutomationControlError e) {
            powerMeterCleanUp();
            throw e;
        }

        this.reactivePowerUpdated = (intValue != this.reactivePower);
        this.eventChanged |= this.reactivePowerUpdated;
        this.reactivePower = intValue;

        this.eventChanged |= forceUpdate;

        panel.unlock();
    }

    public double getTotalKWh() {
        return this.totalKWh / 100.0;
    }

    public String getTotalKWhStr() {
        return fixedPoint(this.totalKWh, 8, 2);
    }

    public double getImportKWh() {
        return this.importKwh / 100.0;
    }

    public String getImportKWhStr() {
        return fixedPoint(this.importKwh, 8, 2);
    }

    public double getExportKWh() {
        return this.exportKWh / 100.0;
    }

    public String getExportKWhStr() {
        return fixedPoint(this.exportKWh, 8, 2);
    }

    public double getActivePower() {
        return this.activePower / 1000.0;
    }

    public String getActivePowerStr() {
        return fixedPoint(this.activePower, 6, 3);
    }

    public double getCurrent() {
        return this.current / 100.0;
    }

    public String getCurrentStr() {
        return fixedPoint(this.current, 4, 2);
    }

    public double getVoltage() {
        return this.voltage / 10.0;
    }

    public String getVoltageStr() {
        return fixedPoint(this.voltage, 4, 1);
    }

    public double getPowerFactor() {
        return this.powerFactor / 1000.0;
    }

    public String getPowerFactorStr() {
        return fixedPoint(this.powerFactor, 4, 3);
    }

    public double getFrequency() {
        return this.frequency / 100.0;
    }

    public String getFrequencyStr() {
        return fixedPoint(this.frequency, 4, 2);
    }

    private int getReactivePowerHelper() throws AutomationControlError {
        final int K_FACTOR = 1000000;
        if (this.powerFactor > 1000 || this.powerFactor < -1000)
            throw new AutomationControlError("Error. Inconsistent data");

        return (int)(this.voltage*this.current*Math.sqrt(K_FACTOR - Math.pow(this.powerFactor, 2))/1000000.0);
    }

    public double getReactivePower() throws AutomationControlError {
        return getReactivePowerHelper() / 1000.0;
    }

    public String getReactivePowerStr() throws AutomationControlError {
        return fixedPoint(getReactivePowerHelper(), 6, 3);
    }

    public Map<String, Object> getSensors(boolean showOnlyUpdated) {
        String key;
        Object value;

        if (this.sensorResult == null)
            this.sensorResult = new TreeMap<>();

        if ((!showOnlyUpdated) && (!this.eventChanged) && (sensorResult.size() != 0))
            return this.sensorResult;

        sensorResult.clear();

        if (showOnlyUpdated) {

            updateHelper(this.sensorResult, KEY_TOTAL_KWH, getTotalKWh(), this.totalKWhUpdated);
            updateHelper(this.sensorResult, KEY_IMPORT_KWH, getImportKWh(), this.importKwhUpdated);
            updateHelper(this.sensorResult, KEY_EXPORT_KWH, getExportKWh(), this.exportKWhUpdated);
            updateHelper(this.sensorResult, KEY_ACTIVE_POWER, getActivePower(), this.activePowerUpdated);

            if (this.reactivePowerUpdated) {
                try {
                    value = getReactivePower();
                    key = KEY_REACTIVE_POWER;
                } catch (AutomationControlError e) {
                    value = e.getMessage();
                    key = KEY_REACTIVE_POWER_FAULT;
                }

                updateHelper(this.sensorResult, key, value, true);
            }

            updateHelper(this.sensorResult, KEY_CURRENT, getCurrent(), this.currentUpdated);
            updateHelper(this.sensorResult, KEY_VOLTAGE, getVoltage(), this.voltageUpdated);
            updateHelper(this.sensorResult, KEY_POWER_FACTOR, getPowerFactor(), this.powerFactorUpdated);
            updateHelper(this.sensorResult, KEY_FREQUENCY, getFrequency(), this.frequencyUpdated);

            return this.sensorResult;
        }

        sensorResult.put(KEY_TOTAL_KWH, getTotalKWh());
        sensorResult.put(KEY_IMPORT_KWH, getImportKWh());
        sensorResult.put(KEY_EXPORT_KWH, getExportKWh());
        sensorResult.put(KEY_ACTIVE_POWER, getActivePower());

        try {
            value = getReactivePower();
            key = KEY_REACTIVE_POWER;
        } catch (AutomationControlError e) {
            value = e.getMessage();
            key = KEY_REACTIVE_POWER_FAULT;
        }

        sensorResult.put(key, value);
        sensorResult.put(KEY_CURRENT, getCurrent());
        sensorResult.put(KEY_VOLTAGE, getVoltage());
        sensorResult.put(KEY_POWER_FACTOR, getPowerFactor());
        sensorResult.put(KEY_FREQUENCY, getFrequency());

        return this.sensorResult;
    }

    public Map<String, Object> getSensorsStr(boolean showOnlyUpdated) {
        String key;
        String value;

        if (this.sensorResultStr == null)
            this.sensorResultStr = new TreeMap<>();

        if ((!showOnlyUpdated) &&(!this.eventChanged) && (sensorResultStr.size() != 0))
            return this.sensorResultStr;

        sensorResultStr.clear();

        if (showOnlyUpdated) {

            updateHelper(this.sensorResultStr, KEY_TOTAL_KWH, getTotalKWhStr(), this.totalKWhUpdated);
            updateHelper(this.sensorResultStr, KEY_IMPORT_KWH, getImportKWhStr(), this.importKwhUpdated);
            updateHelper(this.sensorResultStr, KEY_EXPORT_KWH, getExportKWhStr(), this.exportKWhUpdated);
            updateHelper(this.sensorResultStr, KEY_ACTIVE_POWER, getActivePowerStr(), this.activePowerUpdated);

            if (this.reactivePowerUpdated) {
                try {
                    value = getReactivePowerStr();
                    key = KEY_REACTIVE_POWER;
                } catch (AutomationControlError e) {
                    value = e.getMessage();
                    key = KEY_REACTIVE_POWER_FAULT;
                }

                updateHelper(this.sensorResultStr, key, value, true);
            }

            updateHelper(this.sensorResultStr, KEY_CURRENT, getCurrentStr(), this.currentUpdated);
            updateHelper(this.sensorResultStr, KEY_VOLTAGE, getVoltageStr(), this.voltageUpdated);
            updateHelper(this.sensorResultStr, KEY_POWER_FACTOR, getPowerFactorStr(), this.powerFactorUpdated);
            updateHelper(this.sensorResultStr, KEY_FREQUENCY, getFrequencyStr(), this.frequencyUpdated);

            return this.sensorResultStr;
        }

        sensorResultStr.put(KEY_TOTAL_KWH, getTotalKWhStr());
        sensorResultStr.put(KEY_IMPORT_KWH, getImportKWhStr());
        sensorResultStr.put(KEY_EXPORT_KWH, getExportKWhStr());
        sensorResultStr.put(KEY_ACTIVE_POWER, getActivePowerStr());

        try {
            value = getReactivePowerStr();
            key = KEY_REACTIVE_POWER;
        } catch (AutomationControlError e) {
            value = e.getMessage();
            key = KEY_REACTIVE_POWER_FAULT;
        }

        sensorResultStr.put(key, value);
        sensorResultStr.put(KEY_CURRENT, getCurrentStr());
        sensorResultStr.put(KEY_VOLTAGE, getVoltageStr());
        sensorResultStr.put(KEY_POWER_FACTOR, getPowerFactorStr());
        sensorResultStr.put(KEY_FREQUENCY, getFrequencyStr());

        return this.sensorResultStr;
    }

    public boolean gotEvents() {
        return this.eventChanged;
    }

    public String toString() {
        return this.getSensorsStr(false).toString();
    }
}

File: driver/python/lib/serial/powermeter.py
import math
from lib.automation.registers import Registers
from lib.exception.automationcontrolerror import AutomationControlError
from lib.util.operators import fixedPoint
import panel

class PowerMeter:
    def __init__(self, panel: panel):
        self.panel = panel

    KEY_TOTAL_KWH = "totalKWh"
    KEY_IMPORT_KWH = "importKWh"
    KEY_EXPORT_KWH = "exportKWh"
    KEY_ACTIVE_POWER = "activePower"
    KEY_REACTIVE_POWER = "reactivePower"
    KEY_REACTIVE_POWER_FAULT = "reactivePowerError"
    KEY_CURRENT = "current"
    KEY_VOLTAGE = "voltage"
    KEY_POWER_FACTOR = "powerFactor"
    KEY_FREQUENCY = "frequency"

    totalKWh = 0
    importKwh = 0
    exportKWh = 0
    reactivePower = 0
    activePower = 0
    current = 0
    voltage = 0
    powerFactor = 0
    frequency = 0

    eventChanged = True

    totalKWhUpdated = True
    importKwhUpdated = True
    exportKWhUpdated = True
    reactivePowerUpdated = True
    activePowerUpdated = True
    currentUpdated = True
    voltageUpdated = True
    powerFactorUpdated = True
    frequencyUpdated = True

    sensorResult = {}
    sensorResultStr = {}

    def powerMeterCleanUp(self):
        self.totalKWh = 0
        self.importKwh = 0
        self.exportKWh = 0
        self.reactivePower = 0
        self.activePower = 0
        self.current = 0
        self.voltage = 0
        self.powerFactor = 0
        self.frequency = 0

        self.eventChanged = True

        self.totalKWhUpdated = True
        self.importKwhUpdated = True
        self.exportKWhUpdated = True
        self.reactivePowerUpdated = True
        self.activePowerUpdated = True
        self.currentUpdated = True
        self.voltageUpdated = True
        self.powerFactorUpdated = True
        self.frequencyUpdated = True

    def readSensors(self, forceUpdate = False):
        try:
            intValue = self.panel.modbus.readInt(Registers.Inputs.TOTAL_KWH_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.totalKWhUpdated = (intValue != self.totalKWh)
        self.eventChanged = self.totalKWhUpdated
        self.totalKWh = intValue

        try:
            intValue = self.panel.modbus.readInt(Registers.Inputs.IMPORT_KWH_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.importKwhUpdated = (intValue != self.importKwh)
        self.eventChanged |= self.importKwhUpdated
        self.importKwh = intValue

        try:
            intValue = self.panel.modbus.readInt(Registers.Inputs.EXPORT_KWH_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.exportKWhUpdated = (intValue != self.exportKWh)
        self.eventChanged |= self.exportKWhUpdated
        self.exportKWh = intValue

        try:
            shortValue = self.panel.modbus.readShort(Registers.Inputs.ACTIVE_POWER_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.activePowerUpdated = (shortValue != self.activePower)
        self.eventChanged |= self.activePowerUpdated
        self.activePower = shortValue

        try:
            shortValue = self.panel.modbus.readShort(Registers.Inputs.CURRENT_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.currentUpdated = (shortValue != self.current)
        self.eventChanged |= self.currentUpdated
        self.current = shortValue

        try:
            shortValue = self.panel.modbus.readShort(Registers.Inputs.VOLTAGE_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.voltageUpdated = (shortValue != self.voltage)
        self.eventChanged |= self.voltageUpdated
        self.voltage = shortValue

        try:
            shortValue = self.panel.modbus.readShort(Registers.Inputs.POWER_FACTOR_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.powerFactorUpdated = (shortValue != self.powerFactor)
        self.eventChanged |= self.powerFactorUpdated
        self.powerFactor = shortValue

        try:
            shortValue = self.panel.modbus.readShort(Registers.Inputs.FREQUENCY_REG)
        except:
            self.powerMeterCleanUp()
            raise

        self.frequencyUpdated = (shortValue != self.frequency)
        self.eventChanged |= self.frequencyUpdated
        self.frequency = shortValue

        try:
            intValue = self.getReactivePowerHelper()
        except AutomationControlError as e:
            self.powerMeterCleanUp()
            raise e

        self.reactivePowerUpdated = (intValue != self.reactivePower)
        self.eventChanged |= self.reactivePowerUpdated
        self.reactivePower = intValue

        self.eventChanged |= forceUpdate

    def getTotalKWh(self):
        return self.totalKWh / 100.0

    def getTotalKWhStr(self):
        return fixedPoint(self.totalKWh, 8, 2)

    def getImportKWh(self):
        return self.importKwh / 100.0

    def getImportKWhStr(self):
        return fixedPoint(self.importKwh, 8, 2)

    def getExportKWh(self):
        return self.exportKWh / 100.0

    def getExportKWhStr(self):
        return fixedPoint(self.exportKWh, 8, 2)

    def getActivePower(self):
        return self.activePower / 1000.0

    def getActivePowerStr(self):
        return fixedPoint(self.activePower, 6, 3)

    def getCurrent(self):
        return self.current / 100.0

    def getCurrentStr(self):
        return fixedPoint(self.current, 4, 2)

    def getVoltage(self):
        return self.voltage / 10.0

    def getVoltageStr(self):
        return fixedPoint(self.voltage, 4, 1)

    def getPowerFactor(self):
        return self.powerFactor / 1000.0

    def getPowerFactorStr(self):
        return fixedPoint(self.powerFactor, 4, 3)

    def getFrequency(self):
        return self.frequency / 100.0

    def getFrequencyStr(self):
        return fixedPoint(self.frequency, 4, 2)

    def getReactivePowerHelper(self):
        K_FACTOR = 1000000
        if (self.powerFactor > 1000 or self.powerFactor < -1000):
            raise AutomationControlError("Error. Inconsistent data")
        
        return (int)(self.current*self.voltage*math.sqrt(K_FACTOR - math.pow(self.powerFactor, 2))/1000000.0)

    def getReactivePower(self):
        return self.reactivePower / 1000.0

    def getReactivePowerStr(self):
        return fixedPoint(self.getReactivePowerHelper(), 6, 3)

    def updateHelper(self, destination, key, value, shouldAdd):
        if (shouldAdd):
            destination[key] = value

    def getSensors(self, showOnlyUpdated = False):
        if ((not showOnlyUpdated) and ( not self.eventChanged) and (len(self.sensorResult) != 0)):
            return self.sensorResult

        self.sensorResult.clear()

        if (showOnlyUpdated):
            self.updateHelper(self.sensorResult, self.KEY_TOTAL_KWH, self.getTotalKWh(), self.totalKWhUpdated)
            self.updateHelper(self.sensorResult, self.KEY_IMPORT_KWH, self.getImportKWh(), self.importKwhUpdated)
            self.updateHelper(self.sensorResult, self.KEY_EXPORT_KWH, self.getExportKWh(), self.exportKWhUpdated)
            self.updateHelper(self.sensorResult, self.KEY_ACTIVE_POWER, self.getActivePower(), self.activePowerUpdated)

            if (self.reactivePowerUpdated):
                try:
                    value = self.getReactivePower()
                    key = self.KEY_REACTIVE_POWER
                except AutomationControlError as e:
                    value = e.args[0]
                    key = self.KEY_REACTIVE_POWER_FAULT

                self.updateHelper(self.sensorResult, key, value, True)

            self.updateHelper(self.sensorResult, self.KEY_CURRENT, self.getCurrent(), self.currentUpdated)
            self.updateHelper(self.sensorResult, self.KEY_VOLTAGE, self.getVoltage(), self.voltageUpdated)
            self.updateHelper(self.sensorResult, self.KEY_POWER_FACTOR, self.getPowerFactor(), self.powerFactorUpdated)
            self.updateHelper(self.sensorResult, self.KEY_FREQUENCY, self.getFrequency(), self.frequencyUpdated)

            return self.sensorResult

        self.sensorResult[self.KEY_TOTAL_KWH] = self.getTotalKWh()
        self.sensorResult[self.KEY_IMPORT_KWH] = self.getImportKWh()
        self.sensorResult[self.KEY_EXPORT_KWH] = self.getExportKWh()
        self.sensorResult[self.KEY_ACTIVE_POWER] = self.getActivePower()

        try:
            value = self.getReactivePower()
            key = self.KEY_REACTIVE_POWER
        except AutomationControlError as e:
            value = e.args[0]
            key = self.KEY_REACTIVE_POWER_FAULT

        self.sensorResult[key] = value
        self.sensorResult[self.KEY_CURRENT] = self.getCurrent()
        self.sensorResult[self.KEY_VOLTAGE] = self.getVoltage()
        self.sensorResult[self.KEY_POWER_FACTOR] = self.getPowerFactor()
        self.sensorResult[self.KEY_FREQUENCY] = self.getFrequency()

        return self.sensorResult

    def getSensorsStr(self, showOnlyUpdated = False):
        if ((not showOnlyUpdated) and ( not self.eventChanged) and (len(self.sensorResultStr) != 0)):
            return self.sensorResultStr

        self.sensorResultStr.clear()

        if (showOnlyUpdated):
            self.updateHelper(self.sensorResultStr, self.KEY_TOTAL_KWH, self.getTotalKWhStr(), self.totalKWhUpdated)
            self.updateHelper(self.sensorResultStr, self.KEY_IMPORT_KWH, self.getImportKWhStr(), self.importKwhUpdated)
            self.updateHelper(self.sensorResultStr, self.KEY_EXPORT_KWH, self.getExportKWhStr(), self.exportKWhUpdated)
            self.updateHelper(self.sensorResultStr, self.KEY_ACTIVE_POWER, self.getActivePowerStr(), self.activePowerUpdated)

            if (self.reactivePowerUpdated):
                try:
                    value = self.getReactivePowerStr()
                    key = self.KEY_REACTIVE_POWER
                except AutomationControlError as e:
                    value = e.args[0]
                    key = self.KEY_REACTIVE_POWER_FAULT

                self.updateHelper(self.sensorResultStr, key, value, True)

            self.updateHelper(self.sensorResultStr, self.KEY_CURRENT, self.getCurrentStr(), self.currentUpdated)
            self.updateHelper(self.sensorResultStr, self.KEY_VOLTAGE, self.getVoltageStr(), self.voltageUpdated)
            self.updateHelper(self.sensorResultStr, self.KEY_POWER_FACTOR, self.getPowerFactorStr(), self.powerFactorUpdated)
            self.updateHelper(self.sensorResultStr, self.KEY_FREQUENCY, self.getFrequencyStr(), self.frequencyUpdated)

            return self.sensorResultStr

        self.sensorResultStr[self.KEY_TOTAL_KWH] = self.getTotalKWhStr()
        self.sensorResultStr[self.KEY_IMPORT_KWH] = self.getImportKWhStr()
        self.sensorResultStr[self.KEY_EXPORT_KWH] = self.getExportKWhStr()
        self.sensorResultStr[self.KEY_ACTIVE_POWER] = self.getActivePowerStr()

        try:
            value = self.getReactivePowerStr()
            key = self.KEY_REACTIVE_POWER
        except AutomationControlError as e:
            value = e.args[0]
            key = self.KEY_REACTIVE_POWER_FAULT

        self.sensorResultStr[key] = value
        self.sensorResultStr[self.KEY_CURRENT] = self.getCurrentStr()
        self.sensorResultStr[self.KEY_VOLTAGE] = self.getVoltageStr()
        self.sensorResultStr[self.KEY_POWER_FACTOR] = self.getPowerFactorStr()
        self.sensorResultStr[self.KEY_FREQUENCY] = self.getFrequencyStr()

        return self.sensorResultStr
